### PR TITLE
Add Atomic UIButtonGrid widget

### PR DESCRIPTION
**Pull request by JimMarlowe**
This PR adds an Atomic UIButtonGrid widget. This "widget" is special. It creates a regular grid of buttons, these buttons will remain in a grid organization, even if the container is resized. It will ellipse "..." the text in the buttons if there is not enough room to fully display the text.  This will make it easy to create game menus, and sudoku, tic-tac-toe boards :smile_cat:  There are two methodologies to enter the button text, either by specifying a row and column address, or the "floodfill" add text, which tiles in the text sequentially. The embedded UIButtons can be retrieved by both indexing methodologies, so you can assign event handlers to process the button clicks, or by IDs assigned at creation, using the spreadsheet like scheme, having A1 as the first id, and row incrementing letters, and columns incrementing numbers.
I had to do a little plumbing in Turbobadger to get access to the resize information, otherwise, I'd have to make a TB widget and wrap it.

This is what it can do for the feature examples ...
```
#if ATOMIC_DESKTOP || ATOMIC_MOBILE  
            var buttonGrid = new UIButtonGrid( 6, 4, 6 );
            buttonGrid.SetLayoutHeight( rootLayout.Rect.Height - 32 );
            buttonGrid.SetLayoutWidth( rootLayout.Rect.Width - 32);
            rootLayout.AddChild( buttonGrid );
            var sampleTypes = typeof(Sample).Assembly.GetTypes().Where(t => t.IsSubclassOf(typeof(Sample)) && t != typeof(Sample) ).ToArray();
            foreach (var sample in sampleTypes)
            {
                int got = buttonGrid.AddGridText ( sample.Name );
                var button = buttonGrid.AtGridWidget(got);
                button.SubscribeToEvent<WidgetEvent>( button,  e => 
                {
                    if (e.Type != UI_EVENT_TYPE.UI_EVENT_TYPE_CLICK)  return; // We're only interested in clicks
```
featuredcs

### CODE REVIEW

File: Source/Atomic/UI/UIButtonGrid.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+class ATOMIC_API UIButtonGrid : public UIWidget
+{
+    ATOMIC_OBJECT(UIButtonGrid, UIWidget)
+
+public:
+
+    UIButtonGrid(Context* context, int numRows, int numCols, int margin=0, bool createWidget = true );
+    virtual ~UIButtonGrid();
+
+    virtual String GetGridId( int row, int col ); /// returns id text at r,c, ie, "A1" for 0,0
```
Comment (JoshEngebretson) [inline Source/Atomic/UI/UIButtonGrid.h]: If comments are above the method, they will be included in documentation.  The comments should also be written how you would want them to appear in documentation.  Do these methods all need to be virtual?

File: Source/Atomic/UI/UIButtonGrid.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+   virtual bool OnEvent(const tb::TBWidgetEvent &ev);
+   virtual void OnResized(int old_w, int old_h);
+
+private:
+
+    void GenerateGrid();
+    void ResizeGrid();
+    virtual void AddGridRow( int rownum );
+
+    int m_rows;
```
Comment (JoshEngebretson) [inline Source/Atomic/UI/UIButtonGrid.h]: Member naming convention would be ```rows_``` here, I'd also go with ```columns_``` as it is more clear and matches the full word usage or ```rows_```.  We live in the age of autocomplete :)

**Comment by JimMarlowe** [comment]
Points taken, resolved.

**Comment by JoshEngebretson** [comment]
Awesome, looks great, LANDING! ✈️

**Pull request merged**